### PR TITLE
LM-2713 recurring payouts consumer implementation

### PR DESCRIPTION
**Pull request by waynesayonara**

### CODE REVIEW

File: src/InterestPayout.Common/Application/InterestCalculator.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        /// <list type="bullet">
+        /// <item><description>Target interest payment amount is less than <c>decimal.MinValue</c> or greater than <c>Decimal.MaxValue</c></description></item>
+        /// </list>
+        /// </exception>
+        public static double CalculateInterest(decimal balance, decimal interestRate, int accuracy)
+        {
+            if (balance < decimal.Zero)
+                throw new InvalidOperationException($"Base balance has to be a positive number, but was: '{balance}'");
+            
+            if (interestRate < decimal.MinusOne)
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Application/InterestCalculator.cs]: should be `-100`

File: src/InterestPayout.Common/Application/PayoutConfigService.cs [hunk 2]
```diff
@@ -33,12 +33,18 @@ private static void ValidateConfigs(PayoutConfig[] configs)
...
                 throw new InvalidOperationException(
-                    $"Payouts settings are duplicate for the following assetIds: {string.Join(',', duplicates)}.");
+                    $"Payout settings are duplicate for the following assetIds: {string.Join(',', duplicates)}.");
 
             foreach (var config in configs)
             {
                 if (!Quartz.CronExpression.IsValidExpression(config.PayoutCronSchedule))
                     throw new InvalidOperationException($"Invalid cron expression ('{config.PayoutCronSchedule}') for assetId '{config.AssetId}'.");
+
+                if (config.Accuracy < 0)
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Application/PayoutConfigService.cs]: let's get it from the assets service

File: src/InterestPayout.Common/Application/PayoutConfigService.cs [hunk 3]
```diff
@@ -33,12 +33,18 @@ private static void ValidateConfigs(PayoutConfig[] configs)
...
+                    throw new InvalidOperationException($"Accuracy cannot be negative, but was '{config.Accuracy}' for assetId '{config.AssetId}'.");
+                
+                if (config.PayoutInterestRate < decimal.MinusOne)
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Application/PayoutConfigService.cs]: should be `-100`

File: src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs [hunk 4]
```diff
@@ -1,62 +1,275 @@
...
+                    _logger.LogError("Unexpected duplication of requests to ME {@context}", new
+                    {
+                        IdempotencyId = idempotencyId,
+                        matchingEngineResponse.TransactionId,
+                        matchingEngineResponse.Message,
+                        matchingEngineResponse.Status,
+                        Amount = amount,
+                        command
+                    });
+                    throw new InvalidOperationException(
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs]: we should just log this type of errors

File: src/InterestPayout.Common/Persistence/ReadModels/PayoutSchedules/PayoutScheduleEntity.cs [hunk 5]
```diff
@@ -12,6 +12,8 @@ public class PayoutScheduleEntity
         
         public string CronSchedule { get; set; }
         
+        public int Accuracy { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Persistence/ReadModels/PayoutSchedules/PayoutScheduleEntity.cs]: let's remove it from here and get it from the assets service when needed

File: src/InterestPayout.Common/Application/InterestCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                throw new InvalidOperationException($"Interest rate cannot be less than minus one hundred percent, but was: '{interestRate}'");
+
+            if (accuracy < 0)
+                throw new InvalidOperationException($"Accuracy cannot be negative, but was: '{accuracy}'");
+
+            var decimalAmount = decimal.Multiply(balance, interestRate);
+            var roundedDecimalAmount = decimal.Round(decimalAmount, accuracy, MidpointRounding.ToZero);
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Application/InterestCalculator.cs]: Let's use the same method as here https://github.com/LykkeCity/Lykke.Job.SiriusDepositsDetector/blob/2f68fadc7a00311739f857bd9dc5a9ed6e1abef1/src/Lykke.Job.SiriusDepositsDetector/Services/DepositProcessor.cs#L77

Comment (KonstantinRyazantsev) [reply src/InterestPayout.Common/Application/InterestCalculator.cs]: It's a bit more complicated than just rounding https://github.com/LykkeCity/CommonDotNetLibraries/blob/f315ed300faef84bd61eaca4b62101bcda407528/src/Lykke.Common/Utils.cs#L869

File: src/InterestPayout.Common/Persistence/ReadModels/PayoutSchedules/PayoutScheduleEntity.cs (hunk 5 above)
Comment (waynesayonara) [reply src/InterestPayout.Common/Persistence/ReadModels/PayoutSchedules/PayoutScheduleEntity.cs]: `+`

File: src/InterestPayout.Common/Application/InterestCalculator.cs (hunk 6 above)
Comment (waynesayonara) [reply src/InterestPayout.Common/Application/InterestCalculator.cs]: `+`

File: src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs (hunk 4 above)
Comment (waynesayonara) [reply src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs]: `+`

File: src/InterestPayout.Common/Application/PayoutConfigService.cs (hunk 3 above)
Comment (waynesayonara) [reply src/InterestPayout.Common/Application/PayoutConfigService.cs]: `+`

File: src/InterestPayout.Common/Application/PayoutConfigService.cs (hunk 2 above)
Comment (waynesayonara) [reply src/InterestPayout.Common/Application/PayoutConfigService.cs]: `+`

File: src/InterestPayout.Common/Application/InterestCalculator.cs (hunk 1 above)
Comment (waynesayonara) [reply src/InterestPayout.Common/Application/InterestCalculator.cs]: `+`

File: src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs (hunk 4 above)
Comment (waynesayonara) [reply src/InterestPayout.Worker/Messaging/Consumers/RecurringPayoutCommandConsumer.cs]: `+`

File: src/InterestPayout.Common/Application/InterestCalculator.cs [hunk 7]
```diff
@@ -7,35 +7,68 @@ public static class InterestCalculator
...
 
-            var decimalAmount = decimal.Multiply(balance, interestRate);
-            var roundedDecimalAmount = decimal.Round(decimalAmount, accuracy, MidpointRounding.ToZero);
+            var interest = decimal.Divide(interestRate, 100);
+            var decimalAmount = decimal.Multiply(balance, interest);
+            var roundedDecimalAmount = decimalAmount.TruncateDecimalPlaces(accuracy);
             return Convert.ToDouble(roundedDecimalAmount);
         }
+        
+        public static double TruncateDecimalPlaces(this double value, int places, bool toUpper = false)
```
Comment (KonstantinRyazantsev) [inline src/InterestPayout.Common/Application/InterestCalculator.cs:42]: let's use implementation from Lykke.Common nuget

Comment (KonstantinRyazantsev) [reply src/InterestPayout.Common/Application/InterestCalculator.cs:42]: Decided to do this sepparately together with moving out of direct DB usage of other services

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**